packages/cardano-services/src/Program/utils.ts

```diff
@@ -0,0 +1,98 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { ClientConfig, Pool, QueryConfig } from 'pg';
+import { HttpServerOptions } from './loadHttpServer';
+import { InvalidArgsCombination } from './errors';
+import Logger from 'bunyan';
+import dns from 'dns';
+import pRetry, { FailedAttemptError } from 'p-retry';
+
+export const RETRY_BACKOFF_FACTOR_DEFAULT = 1.1;
+export const RETRY_BACKOFF_MAX_TIMEOUT_DEFAULT = 60 * 1000;
+
+type RetryBackoffConfig = {
+  factor: number;
+  maxRetryTime: number;
+};
+
+export const onFailedAttemptFor =
+  (operation: string, logger: Logger) =>
+  async ({ attemptNumber, message, retriesLeft }: FailedAttemptError) => {
+    const nextAction = retriesLeft > 0 ? 'retrying...' : 'exiting';
+    logger.trace(message);
+    logger.debug(`${operation}: Attempt ${attemptNumber} of ${attemptNumber + retriesLeft}, ${nextAction}`);
+    if (retriesLeft === 0) {
+      logger.error(message);
+      // await server.shutdown(); ?
+      // eslint-disable-next-line unicorn/no-process-exit
+      process.exit(1);
+    }
+  };
+
+export const resolveDnsSrvWithExponentialBackoff = async (
+  hostname: string,
+  { factor, maxRetryTime }: RetryBackoffConfig,
+  logger: Logger
+) =>
+  await pRetry(
+    async () => {
+      // Shall we grab the first one always?
```

Comment:
Is this OK @blaggacao?
``` ts
const [record] = await dns.promises.resolveSrv(hostname);
```

Reply:
A typical DNS SRV query returns a list of all services, eg:
```console
[root@ip-10-52-34-34:~]# dig _infra-node._tcp.service.consul SRV

; <<>> DiG 9.16.27 <<>> _infra-node._tcp.service.consul SRV
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 53825
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 10, AUTHORITY: 0, ADDITIONAL: 41

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;_infra-node._tcp.service.consul. IN	SRV

;; ANSWER SECTION:
_infra-node._tcp.service.consul. 0 IN	SRV	1 1 30911 0a185e5d.addr.eu-central-1.consul.
_infra-node._tcp.service.consul. 0 IN	SRV	1 1 22846 0a185e5d.addr.eu-central-1.consul.
_infra-node._tcp.service.consul. 0 IN	SRV	1 1 27727 0a3479fc.addr.eu-central-1.consul.
_infra-node._tcp.service.consul. 0 IN	SRV	1 1 22324 0a3479fc.addr.eu-central-1.consul.
_infra-node._tcp.service.consul. 0 IN	SRV	1 1 24536 0a18bd22.addr.eu-central-1.consul.
_infra-node._tcp.service.consul. 0 IN	SRV	1 1 27665 0a181be0.addr.eu-central-1.consul.
_infra-node._tcp.service.consul. 0 IN	SRV	1 1 28368 0a185e5d.addr.eu-central-1.consul.
_infra-node._tcp.service.consul. 0 IN	SRV	1 1 26681 0a18bd22.addr.eu-central-1.consul.
_infra-node._tcp.service.consul. 0 IN	SRV	1 1 29224 0a181be0.addr.eu-central-1.consul.
_infra-node._tcp.service.consul. 0 IN	SRV	1 1 31902 0a18bd22.addr.eu-central-1.consul.

;; ADDITIONAL SECTION:
0a185e5d.addr.eu-central-1.consul. 0 IN	A	10.24.94.93
ip-10-24-94-93.eu-central-1.compute.internal.node.eu-central-1.consul. 0 IN TXT	"region=eu-central-1"
ip-10-24-94-93.eu-central-1.compute.internal.node.eu-central-1.consul. 0 IN TXT	"consul-network-segment="
ip-10-24-94-93.eu-central-1.compute.internal.node.eu-central-1.consul. 0 IN TXT	"node_name=client-eu-central-1-c5-2xlarge-infra"
0a185e5d.addr.eu-central-1.consul. 0 IN	A	10.24.94.93
ip-10-24-94-93.eu-central-1.compute.internal.node.eu-central-1.consul. 0 IN TXT	"region=eu-central-1"
ip-10-24-94-93.eu-central-1.compute.internal.node.eu-central-1.consul. 0 IN TXT	"consul-network-segment="
ip-10-24-94-93.eu-central-1.compute.internal.node.eu-central-1.consul. 0 IN TXT	"node_name=client-eu-central-1-c5-2xlarge-infra"
0a3479fc.addr.eu-central-1.consul. 0 IN	A	10.52.121.252
ip-10-52-121-252.us-east-2.compute.internal.node.eu-central-1.consul. 0	IN TXT "region=eu-central-1"
ip-10-52-121-252.us-east-2.compute.internal.node.eu-central-1.consul. 0	IN TXT "consul-network-segment="
ip-10-52-121-252.us-east-2.compute.internal.node.eu-central-1.consul. 0	IN TXT "node_name=client-us-east-2-c5-2xlarge-infra"
0a3479fc.addr.eu-central-1.consul. 0 IN	A	10.52.121.252
ip-10-52-121-252.us-east-2.compute.internal.node.eu-central-1.consul. 0	IN TXT "node_name=client-us-east-2-c5-2xlarge-infra"
ip-10-52-121-252.us-east-2.compute.internal.node.eu-central-1.consul. 0	IN TXT "region=eu-central-1"
ip-10-52-121-252.us-east-2.compute.internal.node.eu-central-1.consul. 0	IN TXT "consul-network-segment="
0a18bd22.addr.eu-central-1.consul. 0 IN	A	10.24.189.34
ip-10-24-189-34.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "node_name=client-eu-central-1-c5-2xlarge-infra"
ip-10-24-189-34.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "region=eu-central-1"
ip-10-24-189-34.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "consul-network-segment="
0a181be0.addr.eu-central-1.consul. 0 IN	A	10.24.27.224
ip-10-24-27-224.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "consul-network-segment="
ip-10-24-27-224.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "node_name=client-eu-central-1-c5-2xlarge-infra"
ip-10-24-27-224.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "region=eu-central-1"
0a185e5d.addr.eu-central-1.consul. 0 IN	A	10.24.94.93
ip-10-24-94-93.eu-central-1.compute.internal.node.eu-central-1.consul. 0 IN TXT	"node_name=client-eu-central-1-c5-2xlarge-infra"
ip-10-24-94-93.eu-central-1.compute.internal.node.eu-central-1.consul. 0 IN TXT	"region=eu-central-1"
ip-10-24-94-93.eu-central-1.compute.internal.node.eu-central-1.consul. 0 IN TXT	"consul-network-segment="
0a18bd22.addr.eu-central-1.consul. 0 IN	A	10.24.189.34
ip-10-24-189-34.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "node_name=client-eu-central-1-c5-2xlarge-infra"
ip-10-24-189-34.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "region=eu-central-1"
ip-10-24-189-34.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "consul-network-segment="
0a181be0.addr.eu-central-1.consul. 0 IN	A	10.24.27.224
ip-10-24-27-224.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "node_name=client-eu-central-1-c5-2xlarge-infra"
ip-10-24-27-224.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "region=eu-central-1"
ip-10-24-27-224.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "consul-network-segment="
0a18bd22.addr.eu-central-1.consul. 0 IN	A	10.24.189.34
ip-10-24-189-34.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "node_name=client-eu-central-1-c5-2xlarge-infra"
ip-10-24-189-34.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "region=eu-central-1"
ip-10-24-189-34.eu-central-1.compute.internal.node.eu-central-1.consul.	0 IN TXT "consul-network-segment="

;; Query time: 209 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Fri Jul 01 05:23:25 UTC 2022
;; MSG SIZE  rcvd: 2123
```

Reply:
I would pick one random and stick with it.  That gives us _some_ very naive load balancing.

Reply:
The proper solution would be a ring hash:
- https://medium.com/swlh/load-balancing-and-consistent-hashing-5fe0156035e1
- https://blog.getambassador.io/load-balancing-strategies-in-kubernetes-l4-round-robin-l7-round-robin-ring-hash-and-more-6a5b81595d6c

Reply:
Ring Hash is a good solution when some "stickiness" is desired, because doing so would increase performance.

Reply:
Side Note: Consul allows us to route the requests in the same data-center only, in case you wondered :smile:

Reply:
It says: `The ring hash approach is used for both “sticky sessions” (where a cookie is set to ensure that all requests from a client arrive at the same pod) and for “session affinity” (which relies on client IP or some other piece of client state).` so should we use some sort of caching the resolved srv record `port` and try to stick with that one in case of on re-resolving or just get the first resolved record would do the job at first stage?

Reply:
Let's just do random selection and make it sticky for **reconnects** by storing a reference in memory. I think it's acceptable since a reconnection timeout indicates there's a complete outage, which is unlikely, and stickiness doesn't really matter in that context.

Reply:
Storing a random selection of resolved DNS address in memory is applied here https://github.com/input-output-hk/cardano-js-sdk/pull/309/commits/b0f80d11b65d8de72b1188aa4c1a33489fb90643

Reply:
I need pair of eyes about the correctness of the applied logic  [getStikyAddressWithDnsSrv](https://github.com/input-output-hk/cardano-js-sdk/pull/309/commits/b9663759a1c458ba742e717761b68bace447ffc6#diff-5935fb07258956001ade74f0edc794b473e06615d67973cc4d416ec157357be3R42-R59)